### PR TITLE
Allow undefined environment variables in config

### LINKED ISSUE #57755: Undefined env config values cause validation warnings
### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/xenodochial-antonelli-9x7v6w?file=%2Fnext.config.js%3A8%2C1

### To Reproduce

1. Run `next dev`
2. See the warnings in the console

### Current vs. Expected behavior

Prior to 13.5.5, an env object with undefined values would not cause any validation warnings. In 13.5.5 (and higher) a message is shown saying
```
⚠ Invalid next.config.js options detected: 
⚠     "env.FOO" is missing, expected string
```

In our case, we set some env vars in the CI environment and leave them out in local development.

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
Binaries:
  Node: 18.17.0
  npm: N/A
  Yarn: N/A
  pnpm: 8.6.12
Relevant Packages:
  next: 13.5.6
  eslint-config-next: 13.5.6
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Additional context

I believe the bug was introduced with https://github.com/vercel/next.js/pull/56383

### PR DESCRIPTION
**Pull request by michaldudak**
This PR loosens the config schema to allow undefined environment variables (for cases where they are defined only on certain environments). Both the Zod schema and TS interface were updated.

Fixes #57755

### CODE REVIEW

File: test/integration/config-validation/test/index.test.ts [hunk 1]
```diff
@@ -54,5 +54,28 @@ describe('next.config.js validation', () => {
...
+      await fs.writeFile(nextConfigPath, configContent)
+      const result = await nextBuild(path.join(__dirname, '../'), undefined, {
+        stderr: true,
+        stdout: true,
+      })
+
+      await fs.remove(nextConfigPath)
+
+      expect(result.stdout + result.stderr).not.toContain(
+        '"env.QUX" is missing, expected string'
```
Comment (timneutkens) [inline test/integration/config-validation/test/index.test.ts:77]: If the expected type ever changes in the future string might not be sufficient but the test wouldn't fail, so changing it to something a bit more generic would catch that case:
```suggestion
        '"env.QUX" is missing'
```

**Comment by michaldudak** [comment]
@timneutkens, can we move this forward?

**Review by timneutkens — APPROVED** [review]
Thanks!

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| buildDuration | 13.1s | 13s | N/A |
| buildDurationCached | 7.3s | 6.3s | N/A |
| nodeModulesSize | 202 MB | 202 MB | ⚠️ +492 B |
| nextStartRea..uration (ms) | 431ms | 409ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 433-HASH.js gzip | 28.3 kB | 28.4 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 240 B | 242 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.5 kB | 98.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| index.html gzip | 527 B | 527 B | ✓ |
| link.html gzip | 540 B | 540 B | ✓ |
| withRouter.html gzip | 523 B | 521 B | N/A |
| Overall change | 1.07 kB | 1.07 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.7 kB | 93.8 kB | N/A |
| page.js gzip | 147 kB | 147 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 623 B | 624 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.4 kB | 37.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | michaldudak/next.js allow-undefined-env-vars | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 168 kB | 168 kB | ✓ |
| app-page-exp..prod.js gzip | 94.2 kB | 94.2 kB | ✓ |
| app-page-tur..prod.js gzip | 94.9 kB | 94.9 kB | ✓ |
| app-page-tur..prod.js gzip | 89.4 kB | 89.4 kB | ✓ |
| app-page.run...dev.js gzip | 138 kB | 138 kB | ✓ |
| app-page.run..prod.js gzip | 88.7 kB | 88.7 kB | ✓ |
| app-route-ex...dev.js gzip | 24.1 kB | 24.1 kB | ✓ |
| app-route-ex..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| app-route.ru...dev.js gzip | 23.5 kB | 23.5 kB | ✓ |
| app-route.ru..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.38 kB | 9.38 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.65 kB | 9.65 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.4 kB | 49.4 kB | ✓ |
| Overall change | 932 kB | 932 kB | ✓ |

</details>

</details>
Commit: 8453059642f10ee23ef63cfdb955883a6c4f7da7

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 8453059642f10ee23ef63cfdb955883a6c4f7da7

`pnpm test-dev test/development/basic/gssp-ssr-change-reloading/test/index.test.ts` 
- GS(S)P Server-Side Change Reloading > should not reload page when client-side is changed too GSSP

<details>
<summary>Expand output</summary>

  ● GS(S)P Server-Side Change Reloading › should not reload page when client-side is changed too GSSP

    expect(received).toEqual(expected) // deep equality

    - Expected  - 1
    + Received  + 1

      Object {
        "count": 1,
        "params": Object {
          "post": "first",
        },
    -   "random": 0.5072538642360738,
    +   "random": 0.8063899941734192,
      }

      220 |
      221 |     const props2 = JSON.parse(await browser.elementByCss('#props').text())
    > 222 |     expect(props).toEqual(props2)
          |                   ^
      223 |
      224 |     await next.patchFile(page, originalContent)
      225 |     await check(() => browser.elementByCss('#change').text(), 'change me')

      at Object.toEqual (development/basic/gssp-ssr-change-reloading/test/index.test.ts:222:19)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by timneutkens** [comment]
Thank you!

**Pull request merged**